### PR TITLE
Add PointStore class

### PR DESCRIPTION
**Pull request by jotok**
* Add bounded, manually managed storage for points.
* Update maven workflow to run on the compact-rcf branch
* Update .gitignore so that the .github directory is not ignored

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+     *
+     * @param index The index value.
+     * @throws IllegalArgumentException if the index value is not valid.
+     * @throws IllegalArgumentException if the current reference count for this
+     *                                  index is nonpositive.
+     */
+    public void decrementRefCount(int index) {
+        checkValidIndex(index);
+
+        // TODO it should be the case that if a ref count is > 0 then size should also
```
Comment (wnbts) [inline Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:142]: minor. this todo can be deleted.

File: Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+     */
+    public int getCapacity() {
+        return capacity;
+    }
+
+    /**
+     * @return the reference count for the given index.
+     */
+    public int getRefCount(int index) {
+        return refCount[index];
```
Comment (wnbts) [inline Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:84]: minor. if the index is not in use, the value can be confusing.

File: Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+    private final double[] store;
+    private final int[] refCount;
+    private final int[] freeBlockStack;
+    private int freeBlockPointer;
+    private final int dimensions;
+    private final int capacity;
+
+    /**
+     * Create a new PointStore with the given block size and capacity.
```
[the remark this answers is not in the document]
Comment (jotok) [reply Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:44]: Fixed.

File: Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+     * @throws IllegalStateException    if the point store is full.
+     */
+    public int add(double[] point) {
+        checkArgument(point.length == dimensions, "point.length must be equal to dimensions");
+        checkState(freeBlockPointer >= 0, "point store is full");
+
+        int nextIndex = freeBlockStack[freeBlockPointer--];
+
+        for (int j = 0; j < dimensions; j++) {
+            store[j + nextIndex * dimensions] = point[j];
```
[the remark this answers is not in the document]
Comment (jotok) [reply Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:109]: Changed.

File: Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+     * Increment the reference count for the given index. This operation assumes
+     * that there is currently a point stored at the given index and will throw an
+     * exception if that's not the case.
+     *
+     * @param index The index value.
+     * @throws IllegalArgumentException if the index value is not valid.
+     * @throws IllegalArgumentException if the current reference count for this
+     *                                  index is nonpositive.
+     */
+    public void incrementRefCount(int index) {
```
[the remark this answers is not in the document]
Comment (jotok) [reply Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:126]: I think the way it currently works is like this: when you go to add a new point to a tree, you traverse to a leaf node. You then compare the point you're adding to the leaf point (but looking up the leaf point values in the PointStore) and if they're the same you increment the counter instead of adding a new point to the store.

If I understand your suggestion, you're saying we could remove the increment method and update the add method so that it checks to see if the point already exists in the store and then increments or inserts accordingly. I think this makes sense if we think of the PointStore as a standalone data structure. In the context of the tree I see the following minor issues:

* During the tree insert step we still need to know if the leaf point is equal to the point being inserted in order to decide whether to create a new parent bounding box. That means that we either need to return the result of the comparison from the add step, or else we need to do the comparison twice (once in the tree insert method and once in the point store add method). The point store add method already returns a value, so it's not as simple as just adding a boolean return value to the add() method.
* We would need a way to easily check for equality in the point store, like a hash map. This will add some memory overhead (and we're trying to be thrifty in this implementation).

To summarize: the current approach is parsimonious and provides exactly the functionality required to use the PointStore in the context of an RCF. Removing the manual increment/decrement operations from the class might be a more convenient API for the PointStore class. We should decide which of these two alternatives we want to prioritize.

File: Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java (hunk 1 above)
Comment (jotok) [reply Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:142]: The two `checkState` calls there are sanity checks and should not actually be reachable states. However, I think we should leave them in while we develop this branch to help catch issues early on, and then remove them later once we're confident that our implementation is correct. I've updated the comment to be clearer about this.

File: Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java (hunk 2 above)
Comment (jotok) [reply Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:84]: I'll update the javadoc for clarity. Currently this method is only used in unit tests, and I think it's convenient to be able to verify that a ref count is 0.

File: Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:126]: Direct answer to Lai's question: we have to keep track of when a point is not being used by any of the trees. This is the flip side of compression.

To amplify Josh's point -- the critical issue is that if we have a point that is a duplicate of a previously seen point, what should be the action? What makes the issue more complex, is that due to the sampling process, it may be that the previous copy was not present in all trees. The solution seems to be that the tree returns some information, which can be a reference (in suitable representation) of the previously seen point. We then need to record that we have an additional copy. This is the increment operation.

Currently, it should be feasible to overload "add" to perform "increment if present" -- if we say "add" a point (float/double array reference, not seen by the pointstore) then the point is explicitly added. If we say "add" an offset, then it performs an increment (after checking validity of offset).

I am not sure about naming conventions, ii.e., how explicitly should names telegraph intent and how much is too much detail in a name. I can see add (double/float[]) and add(int) -- we can also name the second function increment.
I think currently the delete is achieved by "decrement" and if zero occurrences remain, then delete. The major use case over 50 trees will be decrement (and not not delete).

[the remark this answers is not in the document]
Comment (wnbts) [reply Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:126]: Thanks for the answers. My question was more for clarification rather than a suggestion. Here's my thoughts.

Index-based equality check is needed during traversal. In current design, if a new point is a duplicate with the point at the **current index**, reference count can be incremented. That justifies index-based equality and incr.

Then is it possible that a new point is not a duplicate with the point at the current index, for that matter, any points at visited indices, but is a duplicate with some other points at unvisited indices? I think that is possible. Correct?

If the goal is to compress the data within the whole forest, adding such a new point should result into an increment only. In that case, the add method should return a new index for a new point or an existing index for a seen point. Maybe the returned index can also be used for equality check?

Ultimately, i see implementing the equality search a space-time trade-off and hope to make it clear.

[the remark this answers is not in the document]
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:126]: Obviously agree with the larger point: equality search is a space time tradeoff. Good hashes (for the equality check) may require more state (account for collisions), and thereby increase the state. Also the dependencies would build up. One could also use a global RCF tree in addition existing trees to enforce such a check if desired, again that would mean more state.

The intended goal was to have a minimal change and better space usage than the current version. The invariants are: each tree would have exactly one reference for a point (index based). When a new point comes, all trees see the same copy/index. But it is possible that some of the trees had one/alternate previous copy of the point, in which case those trees continue to use their respective copy. One advantage of this that after the initial index generation, trees are decoupled and that may be beneficial if parallelism was turned on.

Maybe focus on this simpler use first and revisit if there is real need for equality check tradeoff?

[the remark this answers is not in the document]
Comment (jotok) [reply Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:126]: I'm open to the idea of doing an equality check when a point is added, but I'm also in favor of merging the current version because we know it works and it will unblock the next PRs. This is an experimental branch, so I think it's OK if we punt on this decision for now.

[the remark this answers is not in the document]
Comment (wnbts) [reply Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:126]: I agree the current implementation is an improvement over the original one. The current choice is clear and reasonable. And further potential trade-off for this part is also clear.
